geo/Geoserver.py

```diff
@@ -430,23 +432,55 @@ def upsert_workspaces_rules(self, workspacePattern: str = '*', permission: str =
 
         created = 0
         updated = 0
+        deleted = 0
+
+        rulesList = self.get_layer_rules()
 
         for workspace in workspaceList:
             try:
-                self.create_layer_rule(workspace['name'], '*', permission, role)
-                created += 1
+                if mode != "D":
+                    self.create_layer_rule(workspace['name'], '*', permission, role)
+                    created += 1
+                else:
+                    error_data = {
+                        'status_code': 409,
+                    }
+                    raise Exception(
+                        "It is not possible to create a rule in Delete mode", error_data)
             except Exception as create_exception:
-                try:
-                    message, error_data = create_exception.args
-                    status_code = error_data['status_code']
-                    if status_code == 409:
-                        self.update_layer_rule(workspace['name'], '*', permission, role)
+                message, error_data = create_exception.args
+                status_code = error_data['status_code']
+                rule = workspace['name'] + '.' + '*' + '.' + permission
+                prevRole = rulesList.get(rule, None)
+                newRole = role
+
+                if prevRole is None:
+                    continue
+
+                if mode == "A":
+                    if prevRole != role:
+                        newRole = ",".join([prevRole, role])
+                elif mode == "D":
+                    if role == prevRole:
+                        self.delete_layer_rule(rule)
+                        deleted += 1
+                        continue
+                    elif role + "," in prevRole:
+                        newRole = prevRole.replace(role + ",", "")
+                    elif "," + role in prevRole:
+                        newRole = prevRole.replace("," + role, "")
+                    else:
+                        continue
+
+                if status_code == 409:
```

Comment:
I know the Replace case is implicit in the definition on line 455, but could we consider assign `newRole` to `role` in an else case here to make it clearer?